### ISSUE TITLE
Change number of spaces in tabs for all generated files

### ISSUE THREAD

**Issue by thardy** [angular/angular-cli#3219]
This was already mentioned in #1252, but since that was ignored/closed without any resolution, I will open a new issue.  The number of tabs angular-cli uses in all generated files needs to be configurable.

It's very frustrating to have to fight our editors whenever we use angular-cli to generate files.  That goes against the entire purpose of the cli as a convenience mechanism.  Most modern editors try hard to honor an existing file's tab spacing, causing angular-cli users to have to fight their editor whenever they generate a new file.  We would also like to be able to keep our editors functioning this way for when we edit other people's source code.  For source code we author ourselves, however, we'd like to control exactly how many spaces are used in tabs without having to fight against our code generation tools.

**Comment by Igonato** [comment]
Overridable/custom templates for `ng generate` would be nice. This way you can at least copy all the standard files and edit them for the desired tabulation. Also being able to make brand new generate snippets would be really cool.

**Comment by Igonato** [comment]
How about making [generate command](https://github.com/angular/angular-cli/blob/master/packages/angular-cli/commands/generate.ts#L23) look in the `project_dir/src/blueprints` first, and only after check blueprints provided by angular-cli?
@filipesilva what do you think about this as a compromise?

**Comment by thardy-pd** [comment]
I like the idea of having a specific folder that angular-cli looks for and, if present, will use any templates present in that folder to override any default templates.  That's what Visual Studio did for most of its T4 templates, and it's a system that works very well.  

Just publish the names of the templates and the default templates themselves, so we can modify them.

Any codegen solution is only as useful as it is customizable.

**Comment by ush189** [comment]
I also would very much like the possibilty to configure the tab size in any way.

**Comment by Brocco** [comment]
@thardy I understand the issue you are having here and it's not a trivial one to solve in a way that will make everyone happy. On the roadmap for the CLI is to provide the ability to have custom blueprints for all generations, so you will be able to have your own blueprints in the future. For right now, we are going to rely on editors/IDEs to handle the re-formatting of generated files so that developers can tailor their environments to how they and their team want their files formatted.

To further explain it is not just a matter of simply specifying the number of tabs or spaces. There are many file formats to consider (i.e. *.ts *.json *.html) that some teams want to mix tabs vs spaces.

I know this is likely not the response/answer you were looking for, but I ask for your patience as we do have a plan to deal with this to ease your frustrations.

I'm going to leave this open for further discussion.

**Comment by thardy** [comment]
I agree that the best and most comprehensive solution is custom blueprints.  That way, everyone can do exactly what they want, when they want.

It's good to hear that's on the roadmap.  Thanks for your response.

**Comment by Equinox** [comment]
@Brocco What happened to keeping it open?

**Comment by Brocco** [comment]
The consensus between the OP (@thardy) and I was sufficient for me to know that this issue is OK to be accommodated currently by IDEs/editors and addressed with custom blueprints in the future.

**Comment by filipesilva** [comment]
This should be available soon-ish via automatically running TS Lint fixers on generators. This is being tracked in #6272.
